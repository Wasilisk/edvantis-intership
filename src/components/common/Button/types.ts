export type ButtonVariants = "filled" | "outlined";
export type ButtonSizes = "small" | "medium" | "large";
export type ButtonHoverVariants = "light" | "dark";

export type ButtonProps = {
    variant: ButtonVariants,
    hoverVariant?: ButtonHoverVariants,
    fullWidth?: boolean,
    size: ButtonSizes,
    label: string
}

export type ButtonStylesProps = Omit<ButtonProps, "label">;
