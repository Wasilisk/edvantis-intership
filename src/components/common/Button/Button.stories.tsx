import {ComponentMeta, Story} from '@storybook/react';
import Button from './index';
import {ButtonProps} from "./types";


export default {
    title: 'Buttons/CommonButton',
    component: Button,
    argTypes: {
        variant: {
            options: ['filled', 'outlined'],
            control: { type: 'radio' },
        },
        size: {
            options: ['small', 'medium', 'large'],
            control: { type: 'select' },
        },
        hoverVariant: {
            options: ['dark', 'light'],
            control: { type: 'select' },
        },
        hasArrow: {
            control: "boolean"
        },
        fullWidth: {
            control: "boolean"
        }
    }
} as ComponentMeta<typeof Button>;

const Template: Story<ButtonProps> = (props) => {
    return <Button {...props}/>;
};

export const OutlinedButton = Template.bind({});
OutlinedButton.args= {
    variant: "outlined",
    label: "Outlined",
    size: "medium",
    fullWidth: false
};

export const FilledButton = Template.bind({});
FilledButton.args= {
    variant: "filled",
    label: "Filled",
    size: "medium",
    hoverVariant: "dark",
    fullWidth: false
};

export const ButtonWithArrow = Template.bind({});
ButtonWithArrow.args= {
    variant: "filled",
    label: "Filled",
    size: "medium",
    hoverVariant: "dark",
    fullWidth: false,
    hasArrow: true
};
