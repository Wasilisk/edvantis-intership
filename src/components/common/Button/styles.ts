import styled, {css} from "styled-components";
import {ButtonHoverVariants, ButtonSizes, ButtonStylesProps, ButtonVariants} from "./types";

const buttonSizeStyles = (size: ButtonSizes) => {
    switch (size) {
        case "small": {
            return css`
              height: 42px;
              font-size: 10px;
              padding: 0 26px;
            `
        }
        case "medium": {
            return css`
              height: 50px;
              font-size: 12px;
              padding: 0 30px;
            `
        }
        case "large": {
            return css`
              height: 55px;
              font-size: 14px;
              padding: 0 36px;
            `
        }
    }
}

const buttonVariantStyles = (variant: ButtonVariants, hoverVariant: ButtonHoverVariants = "dark") => {
    switch (variant) {
        case "outlined": {
            return css`
              border: 2px solid #7bb57c;
              background-color: transparent;
              color: #757171;
              
              :hover {
                background-color: #7bb57c;
                color: white;
              }
            `
        }
        case "filled": {
            return css`
              background-color: #7bb57c;
              border: none;
              color: #ffffff;
              box-shadow: 0 3px 10px 0 rgb(39 37 40 / 28%);
              
              :hover {
                background-color: ${hoverVariant === "dark" ? "#333333" : "#ffffff"};
                color: ${hoverVariant === "light" && "#333333"}
              }
            `
        }
    }
}

export const ButtonStyles = styled.button<ButtonStylesProps>`
  width: ${({fullWidth}) => fullWidth ? "100%" : "fit-content"};
  border-radius: 5px;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  transition: all 0.2s;
  font-family: 'Hind', sans-serif;
  letter-spacing: 0;
  font-weight: 600;
  text-transform: uppercase;
  text-decoration: none;
  
  svg {
    height: 24px;
    width: 24px;
    margin-left: 5px;
  }
  
  :disabled {
    opacity: .5;
    pointer-events: none;
    cursor: not-allowed;
    user-select: none;
  }

  ${({size}) => buttonSizeStyles(size)}
  ${({variant, hoverVariant}) => buttonVariantStyles(variant, hoverVariant)}
`