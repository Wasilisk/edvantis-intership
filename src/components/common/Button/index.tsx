import React from 'react';
import { ButtonStyles } from './styles';
import {ButtonProps} from "./types";
import ButtonArrow from "./ButtonArrow";

const Button = ({label, hasArrow = false, fullWidth = false, hoverVariant = "dark", ...props}: ButtonProps) => {
    return (
        <ButtonStyles fullWidth={fullWidth} hoverVariant={hoverVariant} {...props}>
            {label}
            {
                hasArrow && <ButtonArrow/>
            }
        </ButtonStyles>
    );
};

export default Button;