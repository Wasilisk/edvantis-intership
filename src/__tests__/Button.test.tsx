import {render} from "@testing-library/react";
import Button from "../components/common/Button";

describe('Button component', () => {
    it('should render outlined button', () => {
        const {asFragment} = render(<Button variant="outlined" size="small" label="Click me"/>);
        expect(asFragment()).toMatchSnapshot();
    })
    it('should render filled button with dark hover', () => {
        const {asFragment} = render(<Button variant="filled" size="small" label="Click me"/>);
        expect(asFragment()).toMatchSnapshot();
    })
    it('should render filled button with light hover', () => {
        const {asFragment} = render(<Button variant="filled" size="small" label="Click me" hoverVariant="light"/>);
        expect(asFragment()).toMatchSnapshot();
    })
    it('should render fullwidth button', () => {
        const {asFragment} = render(<Button variant="outlined" size="medium" label="Click me" fullWidth/>);
        expect(asFragment()).toMatchSnapshot();
    })
})