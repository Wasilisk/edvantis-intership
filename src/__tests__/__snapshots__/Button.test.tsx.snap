// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component render filled button with dark hover 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ jnzkTG"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button component render filled button with light hover 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ fpcsXZ"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button component render fullwidth button 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ eqFKKQ"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button component render outlined button 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ hmVncp"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button component renders the component 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ hmVncp"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button component should render button with arrow 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ fAvsjm"
  >
    Click me
    <svg
      height="100%"
      viewBox="0 0 36 36"
      width="100%"
      x="0px"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <path
        class="cls-1"
        d="M34,18l-5-4v3H26v2h3v3ZM24,19H20V17h4v2Zm-6,0H14V17h4v2Zm-6,0H8V17h4v2ZM4,16a2,2,0,1,1-2,2A2,2,0,0,1,4,16Z"
        data-name="28"
        fill="currentColor"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`Button component should render filled button with dark hover 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ jJNmHb"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button component should render filled button with light hover 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ dhZYZA"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button component should render fullwidth button 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ fAvsjm"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button component should render outlined button 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ hdOqBX"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Button component should should call the callback 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ fAvsjm"
  >
    Click me
    <svg
      height="100%"
      viewBox="0 0 36 36"
      width="100%"
      x="0px"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <path
        class="cls-1"
        d="M34,18l-5-4v3H26v2h3v3ZM24,19H20V17h4v2Zm-6,0H14V17h4v2Zm-6,0H8V17h4v2ZM4,16a2,2,0,1,1-2,2A2,2,0,0,1,4,16Z"
        data-name="28"
        fill="currentColor"
      />
    </svg>
  </button>
</DocumentFragment>
`;
